src/protocol/state.rs

```diff
@@ -104,6 +105,34 @@ pub trait ProtocolSim: std::fmt::Debug + Send + Sync + 'static {
         token_out: &Token,
     ) -> Result<GetAmountOutResult, SimulationError>;
 
+    /// Computes the maximum amount that can be traded between two tokens.
+    ///
+    /// This function calculates the maximum possible trade amount between two tokens,
+    /// taking into account the protocol's specific constraints and mechanics.
+    /// The implementation details vary by protocol - for example:
+    /// - For constant product AMMs (like Uniswap V2), this is based on available reserves
+    /// - For concentrated liquidity AMMs (like Uniswap V3), this considers liquidity across tick
+    ///   ranges
+    ///
+    /// Note: if there are no limits, the returned amount will be a "soft" limit,
+    ///       meaning that the actual amount traded could be higher but it's advised to not
+    ///       exceed it.
+    ///
+    /// # Arguments
+    /// * `sell_token` - The address of the token being sold
+    /// * `buy_token` - The address of the token being bought
+    ///
+    /// # Returns
+    /// * `Ok((Option<BigUint>, Option<BigUint>))` - A tuple containing:
+    ///   - First element: The maximum input amount
+    ///   - Second element: The maximum output amount
```

Comment:
I would explain this via domain aka input range for two different functions. `[0, res.0]` is the input domain for `get_amount_out`outside this range the function is "not defined". `[0, res.1]` it is the input domain for `get_amount_in` (we don't expose this yet on this interface).